internal/operator/uninstall.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2020 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operator
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	v1 "github.com/operator-framework/api/pkg/operators/v1"
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/yaml"
+)
+
+type Uninstall struct {
+	config *Configuration
+
+	Package string
+}
+
+func NewUninstall(cfg *Configuration) *Uninstall {
+	return &Uninstall{
+		config: cfg,
+	}
+}
+
+func (u *Uninstall) Run(ctx context.Context) error {
+	subs := v1alpha1.SubscriptionList{}
+	if err := u.config.Client.List(ctx, &subs, client.InNamespace(u.config.Namespace)); err != nil {
+		return fmt.Errorf("list subscriptions: %v", err)
+	}
+
+	var sub *v1alpha1.Subscription
+	for _, s := range subs.Items {
+		if u.Package == s.Spec.Package {
+			sub = &s
+			break
+		}
+	}
+	if sub == nil {
+		return fmt.Errorf("operator package %q not found", u.Package)
+	}
+
+	catsrcKey := types.NamespacedName{
+		Namespace: sub.Spec.CatalogSourceNamespace,
+		Name:      sub.Spec.CatalogSource,
+	}
+	catsrc := &v1alpha1.CatalogSource{}
+	if err := u.config.Client.Get(ctx, catsrcKey, catsrc); err != nil {
+		return fmt.Errorf("get catalog source: %v", err)
+	}
+
+	installPlanKey := types.NamespacedName{
+		Namespace: sub.Status.InstallPlanRef.Namespace,
+		Name:      sub.Status.InstallPlanRef.Name,
+	}
+
+	// Since the install plan is owned by the subscription, we need to
+	// read all of the resource references from the install plan before
+	// deleting the subscription.
+	deleteObjs, err := u.getInstallPlanResources(ctx, installPlanKey)
+	if err != nil {
+		return err
+	}
+
+	// Delete the subscription first, so that no further installs or upgrades
+	// of the operator occur while we're cleaning up.
+	if err := u.config.Client.Delete(ctx, sub); err != nil {
+		return fmt.Errorf("delete subscription %q: %v", sub.Name, err)
+	}
+	fmt.Printf("subscription %q deleted\n", sub.Name)
```

Comment:
There's a mix of print and log statements for status-related messaging in this PR. Can you make them all log statements? This aligns with how `run packagemanifests` outputs status too.

Reply:
Much of this is duplicated from `kubectl-operator`. I'd like to make the `kubectl-operator` functions usable as libraries. In that case, we'll want the logging to be more pluggable. Let me see if I can do that here real quick.

Reply:
Added a `Log func(fmt string, v ...interface{})` field to `Config` and set that to logrus.Infof for cleanup.